samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java

```diff
@@ -167,6 +173,71 @@ public final void onTimer(MessageCollector collector, TaskCoordinator coordinato
     return Collections.emptyList();
   }
 
+  /**
+   * Populate the watermarks based on the following equations:
+   *
+   * <ul>
+   *   <li>InputWatermark(op) = min { OutputWatermark(op') | op1 is upstream of op}</li>
+   *   <li>OutputWatermark(op) = min { InputWatermark(op), OldestWorkTime(op) }</li>
+   * </ul>
+   *
+   * @param watermark incoming watermark
+   * @param collector message collector
+   * @param coordinator task coordinator
+   */
+  public final void onWatermark(Watermark watermark,
+      MessageCollector collector,
+      TaskCoordinator coordinator) {
+    final long inputWatermarkMin;
+    if (prevOperators.isEmpty()) {
+      // for input operator, use the watermark time coming from the source input
+      inputWatermarkMin = watermark.getTimestamp();
+    } else {
+      // InputWatermark(op) = min { OutputWatermark(op') | op1 is upstream of op}
+      inputWatermarkMin = prevOperators.stream().map(op -> op.getOutputWatermarkTime()).min(Long::compare).get();
+    }
+
+    if (inputWatermarkTime < inputWatermarkMin) {
+      // advance the watermark time of this operator
+      inputWatermarkTime = inputWatermarkMin;
+      Watermark inputWatermark = watermark.copyWithTimestamp(inputWatermarkTime);
+      long oldestWorkTime = handleWatermark(inputWatermark, collector, coordinator);
```

Comment:
Right, watermark emission is future task (current pr is for aggregation and propagation). I will discuss with you and Yi about the API to support it.